pkg/ccl/streamingccl/streamingest/cutover_time_processor.go

```diff
@@ -0,0 +1,242 @@
+
+	doneCh        chan struct{}
+	cutoverTimeCh chan hlc.Timestamp
+	wg            ctxgroup.Group
+
+	mu struct {
+		syncutil.Mutex
+
+		pollError error
+	}
+}
+
+// InputWithOutput implements LocalProcessor.
+func (cm *cutoverProcessor) InitWithOutput(
+	ctx context.Context,
+	flowCtx *execinfra.FlowCtx,
+	post *execinfrapb.PostProcessSpec,
+	output execinfra.RowReceiver,
+) error {
+	return cm.Init(
+		ctx,
+		cm,
+		post,
+		streamIngestionCutoverProcessorTypes,
+		flowCtx,
+		0, /* processorID */
+		output,
+		nil, /* memMonitor */
+		execinfra.ProcStateOpts{
+			TrailingMetaCallback: func() []execinfrapb.ProducerMetadata {
+				cm.close()
+				return nil
+			},
+		},
+	)
+
+}
+
+// SetInput implmenents LocalProcessor.
+func (cm *cutoverProcessor) SetInput(ctx context.Context, input execinfra.RowSource) error {
+	return nil
+}
+
+var _ execinfra.LocalProcessor = (*cutoverProcessor)(nil)
+
+// MustBeStreaming implements the execinfra.Processor interface.
+func (cm *cutoverProcessor) MustBeStreaming() bool { return true }
+
+// Start is part of the RowSource interface.
+func (cm *cutoverProcessor) Start(ctx context.Context) {
+	ctx = cm.StartInternal(ctx, "stream-ingestion-cutover-processor")
+
+	cm.wg = ctxgroup.WithContext(ctx)
+	cm.doneCh = make(chan struct{})
+	cm.cutoverTimeCh = make(chan hlc.Timestamp)
+
+	cm.wg.GoCtx(func(ctx context.Context) error {
+		tick := time.NewTicker(cutoverSignalPollInterval.Get(&cm.settings.SV))
+		defer tick.Stop()
+		defer close(cm.cutoverTimeCh)
+
+		var (
+			failureCount        = 0
+			maxRepeatedFailures = 10
+		)
+		for {
+			select {
+			case <-cm.doneCh:
+				return nil
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-tick.C:
+				ts, err := cm.loadCutoverTime(ctx)
+				if err != nil {
+					if failureCount >= maxRepeatedFailures {
+						cm.setPollError(err)
+						return err
+					}
+					failureCount++
+					log.Warningf(ctx, "cutover-time-poller: failed to load cutover time from job: %s (attempt %d/%d)",
+						err.Error(), failureCount, maxRepeatedFailures)
+					continue
+				}
+				failureCount = 0
+				// NOTE: We purposefully always send
+				// the cutover time here, even if it
+				// is empty or the same as a previous
+				// timestamp. Without sending it
+				// peridically, we may never learn
+				// that our consumers are closing.
+				cm.cutoverTimeCh <- ts
+			}
+		}
+	})
+}
+
+// Next is part of the RowSource interface.
+func (cm *cutoverProcessor) Next() (rowenc.EncDatumRow, *execinfrapb.ProducerMetadata) {
+	if cm.State != execinfra.StateRunning {
+		return nil, cm.DrainHelper()
+	}
+
+	select {
+	case cutoverTime, ok := <-cm.cutoverTimeCh:
+		if ok {
+			row, err := encodeCutoverTimeRow(cutoverTime)
+			if err != nil {
+				cm.MoveToDraining(err)
+				return nil, cm.DrainHelper()
+			}
+			return row, nil
+		}
+	case <-cm.Ctx().Done():
+	case <-cm.doneCh:
+		cm.MoveToDraining(nil)
+		return nil, cm.DrainHelper()
+	}
+
+	cm.MoveToDraining(cm.getPollError())
```

Comment:
I believe your reading is correct. When I was writing this, I had convinced myself that this processor stopping because of context.Closed isn't particularly interesting; but, your comment has me reconsidering that. I'll take another look.